service_configuration_lib/cached_view.py

```diff
@@ -0,0 +1,185 @@
+        raise NotImplementedError()
+
+    def remove(self, path: str, service_name: str, config_name: str, config_suffix: str) -> None:
+        raise NotImplementedError()
+
+
+class ConfigsFileWatcher:
+    """This is a cache of the smartstack.yaml config file contents found in
+    /nail/etc/services, with keys of the form '<servicename>.<namespace>'.
+
+    What's special about this cache is that we watch for inotify events on these
+    files, processing these events whenever the `process_events` method is
+    invoked.
+
+    The benefit of the explicit event processing is that it obviates the need
+    for a background thread (which would bring its own set of problems).  The
+    downside is that the `process_events` method must be periodically invoked
+    in order to avoid overflowing the event queue.
+
+    How frequently should we invoked the `process_events` method?  I've measured
+    10 to 20 events per minute, and a queue size (max_queued_events) of 16384.
+    This gives us about half a day of non-processing before the queue overflows
+    and the cache becomes inconsistent.  In the unlikely case that the queue
+    does overflow, then the cache will resetupt itself.
+
+    Args:
+    :param configs_view: implementation of BaseCachedView interface
+    :param configs_folder: Optional, path to configs root folder, `/nail/etc/services`
+    :param services_names: Optional, list of service names to watch
+    :param configs_names: Optional, list of config names to watch
+    :param configs_suffixes: Optional, list of file extentions to watch
+    :param exclude_folders_filters: Optional, filter of masks to exclude folders from watching
+    """
+
+    def __init__(
+        self,
+        configs_view: BaseCachedView,
+        configs_folder: str = DEFAULT_SOA_DIR,
+        services_names: List[str] = ['*'],
+        configs_names: List[str] = ['*'],
+        configs_suffixes: List[str] = ['*'],
+        exclude_folders_filters: List[str] = ['*.~tmp~'],
+    ) -> None:
+        super().__init__()
+        self.configs_view = configs_view
+        self._configs_folder = configs_folder
+        self._services_names = services_names
+        self._configs_names = configs_names
+        self._configs_suffixes = configs_suffixes
+        self._exclude_folders_filters = exclude_folders_filters
+        self.setup()
+
+    @staticmethod
+    def log_inotify_constants() -> None:
+        log.info(f'max_queued_events: {pyinotify.max_queued_events.value}')
+        log.info(f'max_user_instances: {pyinotify.max_user_instances.value}')
+        log.info(f'max_user_watches: {pyinotify.max_user_watches.value}')
+
+    def process_events(self) -> None:
+        """This method should be periodically invoked to process queued inotify
+        events.  See the class docstring for more context.
+        """
+        while self._notifier.check_events(timeout=0):
```

Comment:
It's probably better to limit amount of iterations for the loop, otherwise the current thread could stuck in this function for unspecified amount of time if there are a lot of changes if FS.